### PR TITLE
Add page listing removed beta versions, replacements, and notable changes

### PR DESCRIPTION
**Pull request by liggitt**
xref https://github.com/kubernetes/website/issues/18675

some 1.16 content pulled from https://github.com/kubernetes/website/pull/19014 and https://github.com/kubernetes/website/pull/19388

preview at https://deploy-preview-26483--kubernetes-io-master-staging.netlify.app/docs/reference/using-api/deprecation-guide/

Technical reviews:
- [x] apimachinery - @deads2k 
- [x] auth - @deads2k / @mikedanese / @enj 
- [x] ingress - @robscott

### CODE REVIEW

File: content/en/docs/tasks/administer-cluster/migrate-api/_index.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+Use [client warnings, metrics, and audit information available in 1.19+]https://kubernetes.io/blog/2020/09/03/warnings/#deprecation-warnings
+to locate use of deprecated APIs.
+
+### Migrate to GA APIs
+
+* Update custom integrations and controllers to call the GA APIs
+* Change YAML files to reference the GA APIs
+
+    You can use the `kubectl-convert` command to automatically convert an existing object:
```
Comment (liggitt) [inline content/en/docs/tasks/administer-cluster/migrate-api/_index.md:210]: cc @soltysh - where is this tool available?

**Comment by liggitt** [comment]
/assign @soltysh 
for kubectl-convert bits

/assign @deads2k
for api-machinery and auth bits (APIService, CRD, webhooks)

/assign @robscott 
for ingress/ingressclass bits

/assign @mikedanese @enj 
for csr bits

**Comment by robscott** [comment]
Thanks!

/lgtm

**Comment by liggitt** [comment]
Updates: https://github.com/kubernetes/website/compare/76968446e0bc7d852c88ec52779d1e67371ef031..90048824c16cf9f8ed706d718c53b87c67a39fe4?short_path=22dba43#diff-22dba433fbe16d114c2dcd02ee5916e0462a078fcaa1eba9ea95e9a526a7e6d0

* Led with the version rather than the type to avoid giving people panic attacks
* Clarified the convert command is `kubectl convert` prior to 1.20

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
+---
+reviewers:
+- liggitt
+- lavalamp
+- thockin
+- smarterclayton
+title: "Deprecated API Migration Guide"
+weight: 45
+content_type: api_reference
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:9]: Is this content type right?

We don't have a formal list of content types (and we don't lint against that nonexistent list!) _but_ we might in future start styling API reference pages specially. How about:
```
content_type: reference
```

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+`--runtime-config=$group/$version=false`
+
+For example:
+
+`--runtime-config=admissionregistration.k8s.io/v1beta1=false,apiextensions.k8s.io/v1beta1,...`
+
+### Locate use of deprecated APIs
+
+Use [client warnings, metrics, and audit information available in 1.19+](https://kubernetes.io/blog/2020/09/03/warnings/#deprecation-warnings)
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:247]: This seems like something we should document other than just in a blog article. Putting it in the documentation means it gets tracked for localization and that we try to keep it current.

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+Kubernetes 1.22 will be released later in 2021, so be sure to audit
+your configuration and integrations now!
+
+### Test with deprecated APIs disabled
+
+You can test your clusters by starting an API server with specific API versions disabled
+to simulate upcoming removals. Add the following flag to the API server startup arguments:
+
+`--runtime-config=$group/$version=false`
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:239]: How about:
```suggestion
`--runtime-config=<group>/<version>=false`
```

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+Use [client warnings, metrics, and audit information available in 1.19+](https://kubernetes.io/blog/2020/09/03/warnings/#deprecation-warnings)
+to locate use of deprecated APIs.
+
+### Migrate to non-deprecated APIs
+
+* Update custom integrations and controllers to call the non-deprecated APIs
+* Change YAML files to reference the non-deprecated APIs
+
+    You can use the `kubectl-convert` command (`kubectl convert` prior to v1.20)
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:255]: How do people install `kubectl convert`? I suspect they will want to know (and I don't think we document it).

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    to automatically convert an existing object:
+    
+    `kubectl-convert -f <file> --output-version <group>/<version>`.
+
+    For example, to convert an older Deployment to `apps/v1`, you can run:
+    
+    `kubectl-convert -f ./my-deployment.yaml --output-version apps/v1`
+
+    Note that this may use non-ideal default values. To learn more about a specific
+    resource, check the Kubernetes [api reference](https://kubernetes.io/docs/reference/#api-reference).
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:265]:
```suggestion
    resource, check the Kubernetes [API reference](/docs/reference/kubernetes-api/).
```

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+#### ReplicaSet
+
+The **extensions/v1beta1**, **apps/v1beta1**, and **apps/v1beta2** API versions of ReplicaSet are no longer served as of v1.16.
+
+* Migrate manifests and API clients to use the **apps/v1** API version, available since v1.9.
+* All existing persisted objects are accessible via the new API
+* Notable changes:
+    * `spec.selector` is now required and immutable after creation; use the existing template labels as the selector for seamless upgrades
+
+## What To Do
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:229]: nit:
```suggestion
## What to do
```

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+Kubernetes 1.22 will be released later in 2021, so be sure to audit
+your configuration and integrations now!
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:232]: This is more appropriate for a blog article than for a reference. (This wording is 100% guaranteed to become stale).

If we want to call out the next release we do have shortcodes that let you specify what the next minor release will be and have Hugo update that. There's no shortcode for the year that release happens in though.

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+The **v1.25** release will stop serving the following deprecated API versions:
+
+#### Event
+
+The **events.k8s.io/v1beta1** API version of Event will no longer be served in v1.25.
+
+* Migrate manifests and API clients to use the **events.k8s.io/v1** API version, available since v1.19.
+* All existing persisted objects are accessible via the new API
+* Notable changes
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:34]: What notable changes?

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+* All existing persisted objects are accessible via the new API
+* Notable changes:
+    * `webhooks[*].failurePolicy` default changed from `Ignore` to `Fail` for v1
+    * `webhooks[*].matchPolicy` default changed from `Exact` to `Equivalent` for v1
+    * `webhooks[*].timeoutSeconds` default changed from `30s` to `10s` for v1
+    * `webhooks[*].sideEffects` default value is removed, and the field made required, and only `None` and `NoneOnDryRun` are permitted for v1
+    * `webhooks[*].admissionReviewVersions` default value is removed and the field made required for v1 (supported versions for AdmissionReview are `v1` and `v1beta1`)
+    * `webhooks[*].name` must be unique in the list for objects created via `admissionregistration.k8s.io/v1`
+
+#### CustomResourceDefinitions
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:62]: nit:
```suggestion
#### CustomResourceDefinition
```

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+
+#### IngressClass
+
+The **networking.k8s.io/v1beta1** API version of IngressClass will no longer be served in v1.22.
+
+* Migrate manifests and API clients to use the **networking.k8s.io/v1** API version, available since v1.19.
+* All existing persisted objects are accessible via the new API
+* No notable changes
+
+#### RBAC
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:147]: To make it clear that this isn't talking about a RBAC resource where you could `kubectl get rbac`, I recommend grouping them as:
```suggestion
#### RBAC resources
```

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+#### PriorityClass
+
+The **scheduling.k8s.io/v1beta1** API version of PriorityClass will no longer be served in v1.22.
+
+* Migrate manifests and API clients to use the **scheduling.k8s.io/v1** API version, available since v1.14.
+* All existing persisted objects are accessible via the new API
+* No notable changes
+
+#### Storage
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:163]: To make it clear that this isn't talking about a resource where you could `kubectl get storage`, I recommend grouping them as:
```suggestion
#### Storage resources
```

File: content/en/docs/reference/using-api/deprecation-guide.md [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+
+The **storage.k8s.io/v1beta1** API version of CSIDriver, CSINode, StorageClass, and VolumeAttachment will no longer be served in v1.22.
+
+* Migrate manifests and API clients to use the **storage.k8s.io/v1** API version
+  * CSIDriver is available in **storage.k8s.io/v1** since v1.19.
+  * CSINode is available in **storage.k8s.io/v1** since v1.17
+  * StorageClass is available in **storage.k8s.io/v1** since v1.6
+  * VolumeAttachment is available in **storage.k8s.io/v1** v1.13
+* All existing persisted objects are accessible via the new API
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-guide.md:172]: nit:
```suggestion
* All existing persisted objects are accessible via the new APIs
```

**Review by sftim — COMMENTED** [review]
I think this is 
/sig release

does that seem right?

- For all the API headings, I would write custom fragment identfiers.
  For example,
  ```diff
  - #### RuntimeClasseprecation}
  ```
  so that if we ever need to deprecate two APIs that have the same `kind` in different release versions, we've left room for that.
- I recommend hyperlinking to the API reference page for each resource that is mentioned.
- I recommend editing https://kubernetes.io/docs/contribute/style/style-guide/ to document that this page is an exception from [_avoid statements about the future_](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future).
- I also added inline queries and suggestions.

**Comment by sftim** [comment]
Looking at https://github.com/kubernetes/website/issues/18675#issuecomment-574213524 then I think the more we can make this look like generated output, the easier it becomes to switch over to generation.

**Comment by liggitt** [comment]
Thanks for the detailed review! Will fix up shortly

File: content/en/docs/reference/using-api/deprecation-guide.md (hunk 5 above)
Comment (sftim) [reply content/en/docs/reference/using-api/deprecation-guide.md:255]: (it's OK, IMO, to log an issue about documenting it and making it clear there's an open issue on that topic - maybe we mark it insta-frozen so it doesn't get overlooked).

File: content/en/docs/reference/using-api/deprecation-guide.md (hunk 3 above)
Comment (liggitt) [reply content/en/docs/reference/using-api/deprecation-guide.md:247]: +1 to having a dedicated docs page for it. opened https://github.com/kubernetes/website/issues/26535 with a pointer back to this PR to update the link once the dedicated page is available

File: content/en/docs/reference/using-api/deprecation-guide.md (hunk 5 above)
Comment (liggitt) [reply content/en/docs/reference/using-api/deprecation-guide.md:255]: `kubectl convert` prior to 1.20 doesn't need installing. https://github.com/kubernetes/kubernetes/issues/99076 tracks making a pre-built `kubectl-convert` easily available/installable.

File: content/en/docs/reference/using-api/deprecation-guide.md (hunk 8 above)
Comment (liggitt) [reply content/en/docs/reference/using-api/deprecation-guide.md:232]: dropped the immediately stale text, left the specific tasks

File: content/en/docs/reference/using-api/deprecation-guide.md (hunk 9 above)
Comment (liggitt) [reply content/en/docs/reference/using-api/deprecation-guide.md:34]: oops, added

**Comment by liggitt** [comment]
> * I recommend hyperlinking to the API reference page for each resource that is mentioned.

are the links stable? from my feedback on https://github.com/kubernetes/website/issues/25505#issuecomment-763656285, I'm not sure they will be

> The URLs include the categorizations (".../services-resources/...", ".../workloads-resources/...", etc) which are somewhat arbitrary. Making the URLs based on the API group/version/kind will produce more stable URLs

> Looking at [#18675 (comment)](https://github.com/kubernetes/website/issues/18675#issuecomment-574213524) then I think the more we can make this look like generated output, the easier it becomes to switch over to generation.

I agree, though I don't have a great data source for the "notable changes" bit at the moment. Still, I tried to keep the format identical for each type so that once we do have a good data source for that, we could approximate this structure with a generator.

**Comment by liggitt** [comment]
> I recommend editing https://kubernetes.io/docs/contribute/style/style-guide/ to document that this page is an exception from avoid statements about the future.

done

**Comment by liggitt** [comment]
inline comments accepted or tracking issues created, I think the only open question is about whether the hyperlinks to the API reference pages are stable enough to link all of these types to

**Comment by sftim** [comment]
> the hyperlinks to the API reference pages are stable enough to link all of these types to

That is a big question! Kind of already covered in https://github.com/kubernetes/website/issues/25505

**Comment by liggitt** [comment]
> That is a big question! Kind of already covered in #25505
> 
> I mean, #25505 _asks_ that question.

Yeah... I'm happy to link to stable reference docs once we have them, but it's not clear the new API reference URLs are what we want long-term.

**Comment by liggitt** [comment]
cc @kubernetes/api-reviewers 
/assign @smarterclayton 
/sig architecture 
For overall page review

**Comment by dims** [comment]
/approve
/lgtm

from sig-arch side!

**Comment by sftim** [comment]
/hold
to wait for @k/api-reviewers LGTM (feel free to unhold if you're part of the API review subproject and are adding LGTM).

/approve

**Comment by sftim** [comment]
I think this is also going to prompt an issue to mention this page in the SIG Docs part of the release process.

File: content/en/docs/tasks/administer-cluster/migrate-api/_index.md (hunk 1 above)
Comment (soltysh) [reply content/en/docs/tasks/administer-cluster/migrate-api/_index.md]: Fixed in https://github.com/kubernetes/kubernetes/pull/99155

**Comment by smarterclayton** [comment]
/lgtm
/hold cancel

from API review

File: content/en/docs/reference/using-api/deprecation-guide.md (hunk 5 above)
Comment (sftim) [reply content/en/docs/reference/using-api/deprecation-guide.md:260]: We never logged the docs issue as far as I know. This came up: https://github.com/kubernetes/website/issues/38809

**Pull request merged**